add model to investigate associations with 3rd dose

elsie-h opened the pull request:
As suggested by reviewers, I've added a model to investigate associations between patient characteristics and third dose.

The attached word doc outlines specifications for the model (feel free to make edits and upload in comments).

[waning-1year-additional-analyses.docx](https://github.com/opensafely/waning-ve-2dose-1year/files/11565856/waning-1year-additional-analyses.docx)

Please review the following three files:

- [analysis/study_definition_prop.py](https://github.com/opensafely/waning-ve-2dose-1year/blob/aje-revision-1/analysis/study_definition_prop.py)
- [analysis/propensity/data_propmodel.R](https://github.com/opensafely/waning-ve-2dose-1year/blob/aje-revision-1/analysis/propensity/data_propmodel.R)
- [analysis/propensity/fit_propmodel.R](https://github.com/opensafely/waning-ve-2dose-1year/blob/aje-revision-1/analysis/propensity/fit_propmodel.R)

elsie-h committed:
Create study_definition_prop.py

elsie-h committed:
Create propensity_dose3.R

elsie-h committed:
update lib data

elsie-h committed:
add generate_prop_data to project.yaml

elsie-h committed:
Update propensity_dose3.R

elsie-h committed:
Update propensity_dose3.R

elsie-h committed:
Update propensity_dose3.R

elsie-h committed:
Update propensity_dose3.R

elsie-h committed:
Delete propensity_dose3.R

elsie-h committed:
Create data_propmodel.R

elsie-h committed:
Create fit_propmodel.R

elsie-h committed:
Create postprocess_propmodel.R

elsie-h committed:
Update dummy_data_vax.feather

elsie-h committed:
update project.yaml

elsie-h committed:
Delete postprocess_propmodel.R

elsie-h committed:
Update data_propmodel.R

elsie-h committed:
Update data_propmodel.R

elsie-h committed:
Update fit_propmodel.R

elsie-h committed:
update project yaml

wjchulme reviewed (approved):
Just a few of comments

* maybe need to add cancer to model (as discussed) or not, if mostly wrapped up in JCVI groups
* the model itself is fine, though there may be too many strata? I would argue maybe not worth including region, as it will be much less influential in vaccine uptake than it is in covid outcomes, and will significantly increase run-time.
* The wrangling of time-varying data could be clearer! But I think it's doing what you want it to do (difficult without sensible dummy data to check). In the booster effectiveness repo (now been superseded by your work), we processed each time-varying variable separately, combined them into a long one-row-per-event dataset, and then put it all in a Cox-compatible dataset using `tmerge`  https://github.com/opensafely/booster-effectiveness/blob/main/analysis/data_process_long.R. This dealt with the potentially overlapping hospitalisations neatly as it just ended up as a status indicator over time which dealt naturally with issues 1,2,3 that you mention. Pointing this out in case it helps, either just with a small test to confirm what you've done is sensible or to fully refactor in this way for clarify. But up to you

elsie-h committed:
add cancer codelists

elsie-h committed:
add cancer variables

elsie-h committed:
remove region from cox model strata

elsie-h committed:
update default expectations

elsie-h committed:
simplify derivation of data_timevarying

elsie-h committed:
Update fit_propmodel.R

elsie-h committed:
reorganise files

elsie-h committed:
update annotations

elsie-h commented:
Thanks @wjchulme 

NB: I've renamed the files to make it more obvious what they do, so the diffs are not very helpful but I outline the changes below.

## Add cancer to model:

Cancers and cancer therapies that result in immunosuppression were already included [in the `immunosuppressed` baseline variable](https://github.com/opensafely/waning-ve-2dose-1year/blob/32ebb38a3736de448a5bd2e8c1883dc8b7fb6174/analysis/study_definition_covs.py#L256-L272), but in the dose3 model we're interested in any indicators of poor health so I agree we should add cancer to the model.

For now, I've only [defined cancer at baseline](https://github.com/opensafely/waning-ve-2dose-1year/blob/32ebb38a3736de448a5bd2e8c1883dc8b7fb6174/analysis/dose3/data_timevarying.R#L69-L81), as this is consistent with the other disease variables that are only defined at baseline. However, I've [extracted the data to update it after baseline](https://github.com/opensafely/waning-ve-2dose-1year/blob/32ebb38a3736de448a5bd2e8c1883dc8b7fb6174/analysis/study_definition_timevarying.py#L164-L195), in case we decide to do this.

## Remove region from Cox strata:
Agreed - updated formula [here](https://github.com/opensafely/waning-ve-2dose-1year/blob/32ebb38a3736de448a5bd2e8c1883dc8b7fb6174/analysis/dose3/dose3_model.R#L59-L63).

## Wrangling of time-varying data:
Thanks for linking your script - I've re-written this using `tmerge` and it's much more simple now!! [Here's the updated script](https://github.com/opensafely/waning-ve-2dose-1year/blob/aje-revision-1/analysis/dose3/data_timevarying.R).

Just one modification:

- I've transformed all variables at once to long format rather than one at a time as in your script ([lines here](https://github.com/opensafely/waning-ve-2dose-1year/blob/f08a9bab928a29d4a24162422f3e57acc1045db1/analysis/dose3/data_timevarying.R#L85-L108))

elsie-h committed:
use tdcstart option

wjchulme commented:
Looks good!

elsie-h committed:
correct typo in cancer_0_date

elsie-h committed:
various updates to dose3_model

update object names prop->3dose, add data summaries, correct typo commenting out cancer from formula

Update dose3_model.R

elsie-h committed:
update error printing for dose3_model
